### PR TITLE
[SPARK-44533][PYTHON] Add support for accumulator, broadcast, and Spark files in Python UDTF's analyze

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Adds support for `accumulator`, `broadcast` in vanilla PySpark, and Spark files in both vanilla PySpark and Spark Connect Python client, in Python UDTF's analyze.

For example, in vanilla PySpark:

```py
>>> colname = sc.broadcast("col1")
>>> test_accum = sc.accumulator(0)

>>> @udtf
... class TestUDTF:
...     @staticmethod
...     def analyze(a: AnalyzeArgument) -> AnalyzeResult:
...         test_accum.add(1)
...         return AnalyzeResult(StructType().add(colname.value, a.data_type))
...     def eval(self, a):
...         test_accum.add(1)
...         yield a,
...
>>> df = TestUDTF(lit(10))
>>> df.printSchema()
root
 |-- col1: integer (nullable = true)

>>> df.show()
+----+
|col1|
+----+
|  10|
+----+

>>> test_accum.value
2
```

or

```py
>>> pyfile_path = "my_pyfile.py"
>>> with open(pyfile_path, "w") as f:
...     f.write("my_func = lambda: 'col1'")
...
24
>>> sc.addPyFile(pyfile_path)
>>> # or spark.addArtifacts(pyfile_path, pyfile=True)
>>>
>>> @udtf
... class TestUDTF:
...     @staticmethod
...     def analyze(a: AnalyzeArgument) -> AnalyzeResult:
...         import my_pyfile
...         return AnalyzeResult(StructType().add(my_pyfile.my_func(), a.data_type))
...     def eval(self, a):
...         yield a,
...
>>> df = TestUDTF(lit(10))
>>> df.printSchema()
root
 |-- col1: integer (nullable = true)

>>> df.show()
+----+
|col1|
+----+
|  10|
+----+
```

### Why are the changes needed?

To support missing features: `accumulator`, `broadcast`, and Spark files in Python UDTF's analyze.

### Does this PR introduce _any_ user-facing change?

Yes, accumulator, broadcast in vanilla PySpark, and Spark files in both vanilla PySpark and Spark Connect Python client will be available.

### How was this patch tested?

Added related tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @allisonwang-db @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/tests/connect/test_parity_udtf.py [hunk 1]
```diff
@@ -104,6 +106,23 @@ def terminate(self):
...
 
+    @unittest.skip("Spark Connect does not support broadcast but the test depends on it.")
+    def test_udtf_with_analyze_using_broadcast(self):
+        super().test_udtf_with_analyze_using_broadcast()
+
+    @unittest.skip("Spark Connect does not support accumulator but the test depends on it.")
+    def test_udtf_with_analyze_using_accumulator(self):
+        super().test_udtf_with_analyze_using_accumulator()
+
+    def _add_pyfile(self, path):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_parity_udtf.py:117]: Nice, I like it.

File: python/pyspark/sql/tests/test_udtf.py [hunk 2]
```diff
@@ -1214,6 +1217,153 @@ def eval(self, **kwargs):
...
+            @staticmethod
+            def analyze(a: AnalyzeArgument) -> AnalyzeResult:
+                test_accum.add(1)
+                return AnalyzeResult(StructType().add("col1", a.data_type))
+
+            def eval(self, a):
+                test_accum.add(1)
+                yield a,
+
+        df = TestUDTF(lit(10))
```
Comment (allisonwang-db) [inline python/pyspark/sql/tests/test_udtf.py:1250]: Just curious, does the accumulator work if we use the UDTF in a SQL query?

File: python/pyspark/sql/tests/test_udtf.py [hunk 3]
```diff
@@ -1214,6 +1217,153 @@ def eval(self, **kwargs):
...
+        assertSchemaEqual(df.schema, StructType().add("col1", IntegerType()))
+        assertDataFrameEqual(df, [Row(col=10)])
+        self.assertEqual(test_accum.value, 2)
+
+    def _add_pyfile(self, path):
+        self.sc.addPyFile(path)
+
+    def test_udtf_with_analyze_using_pyfile(self):
```
Comment (allisonwang-db) [inline python/pyspark/sql/tests/test_udtf.py:1258]: This is really cool. Can we add/modify this test case to use pyfile in 1) eval 2) terminate?

**Review by allisonwang-db — COMMENTED** [review]
This is great!

File: python/pyspark/sql/tests/test_udtf.py (hunk 2 above)
Comment (ueshin) [reply python/pyspark/sql/tests/test_udtf.py:1250]: Yes, let me update the test.

**Review by allisonwang-db — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**